contracts/oracle/src/contract/oracle/mod.rs

```diff
@@ -155,3 +155,137 @@ where
+
+    type BaseCurrency = Usdc;
+
+    type NlsCoin = Coin<Nls>;
+    type UsdcCoin = Coin<Usdc>;
+
+    const NOW: Timestamp = Timestamp::from_seconds(1);
+
+    #[test]
+    fn test() {
+        let mut storage: MockStorage = MockStorage::new();
+
+        let price_config: PriceConfig = PriceConfig::new(
+            Percent::HUNDRED,
+            Duration::from_secs(1),
+            1,
+            Percent::HUNDRED,
+        );
+
+        init(&mut storage, &price_config);
+
+        add_alarm(&mut storage);
+
+        feed_price(&price_config, &mut storage);
+
+        dispatch_and_deliver(&mut storage, 1);
+
+        // Bug happens on this step.
+        dispatch_and_deliver(&mut storage, 0);
+    }
+
+    #[track_caller]
+    fn init(storage: &mut dyn Storage, price_config: &PriceConfig) {
+        Feeders::try_register(
+            DepsMut {
+                storage,
+                api: &MockApi::default(),
+                querier: QuerierWrapper::new(&MockQuerier::<Empty>::new(&[])),
+            },
+            String::from("feeder"),
+        )
+        .unwrap();
+
+        Config::new(String::from(BaseCurrency::TICKER), price_config.clone())
+            .store(storage)
+            .unwrap();
+
+        SupportedPairs::<BaseCurrency>::new(
+            swap_tree!({ base: Usdc::TICKER }, (1, Nls::TICKER)).into_tree(),
+        )
+        .unwrap()
+        .save(storage)
+        .unwrap();
+    }
+
+    #[track_caller]
+    fn add_alarm(storage: &mut dyn Storage) {
+        let storage: &mut dyn Storage = storage;
+
+        let mut alarms: MarketAlarms<'_, &mut dyn Storage> = MarketAlarms::new(storage);
+
+        alarms
+            .try_add_price_alarm::<BaseCurrency>(
+                Addr::unchecked("1"),
+                Alarm::new(
+                    SpotPrice::new(NlsCoin::new(1).into(), UsdcCoin::new(1).into()),
+                    Some(SpotPrice::new(
+                        NlsCoin::new(1).into(),
+                        UsdcCoin::new(2).into(),
+                    )),
+                ),
+            )
+            .unwrap();
+    }
+
+    #[track_caller]
+    fn feed_price(price_config: &PriceConfig, storage: &mut dyn Storage) {
+        Feeds::<BaseCurrency>::with(price_config.clone())
+            .feed_prices(
+                storage,
+                NOW,
+                &Addr::unchecked("feeder"),
+                &[price::total_of(NlsCoin::new(1)).is(UsdcCoin::new(2)).into()],
+            )
+            .unwrap();
+    }
+
+    #[track_caller]
+    fn dispatch(storage: &mut dyn Storage, expected_count: u32) {
+        let mut oracle: Oracle<'_, &mut dyn Storage, _> =
+            Oracle::<'_, _, BaseCurrency>::load(storage).unwrap();
+
+        let alarms: u32 = oracle.try_notify_alarms(NOW, 16).unwrap().0;
+
+        assert_eq!(alarms, expected_count);
+    }
+
+    #[track_caller]
+    fn deliver(storage: &mut dyn Storage, count: u32) {
+        let storage: &mut dyn Storage = storage;
```

Comment:
???

Reply:
Leftover from the minimization. Removing it now.